src/main/scala/asciibot/parser/parser.scala

```diff
@@ -0,0 +1,200 @@
+    val nlns = lns map (_.padTo(maxLen, ' '))
+
+    val rules: MutableList[Rule] = new MutableList
+
+    var ruleStart = 0
+
+    // iterate over columns
+    var i = 0
+    for (i <- 0 until maxLen) {
+      // Number of bars in a colum
+      val numBars = (nlns map (l => l(i)) filter (_=='|')).length
+      if (!(numBars == 0 || numBars == 3)) {
+        //throw new RuleSeparatorError(i)
+        throw new CharBlockError("Unaligned vertical bars", i,i+1)
+      }
+      if (numBars == 3) {
+        rules += parseRule(nlns, ruleStart, i)
+        ruleStart = i + 1
+      }
+    }
+    rules += parseRule(nlns, ruleStart, maxLen)
+
+    rules.toList
+  }
+
+  def parseRule(lines: Seq[String], start: Int, end: Int) : Rule = {
+    def throwBE(m : String) = throw new CharBlockError(m, start, end)
+    val arrowStart = lines(1).indexOfSlice("->", start)
+    if (arrowStart < start || arrowStart >= end) {
+      throwBE("No arrow found")
+    }
+    val (inStateStart, inStateEnd) = trimBlock(lines, start, arrowStart)
+    val (outStateStart, outStateEnd) = trimBlock(lines, arrowStart + 2, end) 
+    val (surrs, inState) = parseInState(lines, inStateStart, inStateEnd)
+    val (md, outState) = parseOutState(lines, outStateStart, outStateEnd, surrs)
+    Rule(inState, surrs, md, outState)
+  }
+
+  def trimBlock(lines: Seq[String], start: Int, end: Int) : (Int, Int) = {
+    var s = start
+    var e = end
+    while (lines(0)(s) == ' ' && lines(1)(s) == ' ' && lines(2)(s) == ' ') {
+      s += 1;
+    }
+    while (lines(0)(e - 1) == ' ' && lines(1)(e - 1) == ' ' && lines(1)(e - 1) == ' ') {
+      e -= 1;
+    }
+    (s, e)
+  }
+
+  def parseInState(a: Seq[String], start: Int, end: Int) : (Surroundings, State) = {
+    def throwBE(m : String) = throw new CharBlockError(m, start, end)
+    assert(a.length == 3)
+    val state_str = a(1).slice(start+1,end-1)
+    // Check to make sure there are no illegal characters in the state name
+    if ("#*_|-/> " exists (state_str contains _)) {
+      //throw new StateNameError(state_str)
+      throw new CharBlockError(s"Bad state name `$state_str'", start, end)
+    }
+
+    if (   a(0)(start) != ' '
+        || a(0)(end-1) != ' '
+        || a(2)(start) != ' '
+        || a(2)(end-1) != ' ') {
+        throwBE("Corners of state box not empty")
+    }
+
+    // NEWS order, strings
+    val str_surrs = List(
+      ("Top", a(0).slice(start,end) filter (_ != ' ')), // TODO: change to partition or something
+      ("Right", a(1)(end-1).toString),
+      ("Left", a(1)(start).toString),
+      ("Bottom", a(2).slice(start,end)  filter (_ != ' '))
+    )
+
+    val surrs = str_surrs map (_ match {
+        case (_, "#") => Blocked
+        case (_, "_") => Open
+        case (_, "*") => Anything
+        case (dir, s) => throw new CharBlockError(s"Invalid Surrounding `$s' for $dir side", start, end)
+      })
+
+    (Surroundings(surrs(0), surrs(1), surrs(2), surrs(3)), State(state_str))
```

Comment:
This seems like a violation of the Single Responsibility Principle (which isn't necessarily bad - they're guidelines, not rules) but is there a way you could separate it out more cleanly?